Add scope/namespace option to CodeArtifact login

ryansonshine opened the pull request:
Resolves #5291 

*Description of changes:*

This change adds the ability to pass in a scope/namespace to CodeArtifact login.
The primary arg is `namespace` to remain generalized across different CodeArtifact
tools (ie: npm scope, maven groupid).

This change also includes:

- Throw error if namespace or scope are used for configuring pip
- Throw error if namespace or scope are used for configuring twine
- Throw error if namespace and scope are used at the same time
- Accept npm namespace value with or without "@" prefix for scope
- Add "@" prefix to npm scope if user has not passed it
- Validation on scope name to adhere with npm rules


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

kyleknap commented on tests/unit/customizations/codeartifact/test_adapter_login.py:
```diff
@@ -63,7 +63,6 @@ def test_run_commands_unhandled_error(self):
         with self.assertRaisesRegexp(OSError, 'unhandled error'):
             self.test_subject._run_commands(tool, ['echo', tool])
 
-
```
Let's keep the newline here to follow pep8 practices of two newlines between classes (This happens again later on in the diff as well). In general, it would not hurt to run [`flake8`](https://flake8.pycqa.org/en/latest/) on the files being updated to make sure they follow pep8 style.

ryansonshine replied:
Will do! Thanks

kyleknap commented on awscli/customizations/codeartifact/login.py:
```diff
@@ -314,6 +348,17 @@ class CodeArtifactLogin(BasicCommand):
                          'domain',
             'required': False,
         },
+        {
+            'name': 'scope',
+            'help_text': 'Associates a scope with your repository for npm '
+                         '(alias for namespace)',
+            'required': False,
+        },
```
I'm wondering if we need both of the `--scope` and `--namespace` parameters? And instead we just expose the `--namespace` parameter and build up the documentation around how it applies to npm sort of like how the CodeArtifact docs describe namespaces: https://docs.aws.amazon.com/codeartifact/latest/ug/codeartifact-concepts.html#welcome-concepts-package-namespace.

The main reason that I'm asking/suggesting it is because it is a bit confusing that there are two parameters that are essentially aliases but one of them would only ever apply to a particular tool. Furthermore, in the future we can always add the `--scope` alias (or expose it in a different way that javascript developers would expect) back if `--namespace` is too general so that gives us a bit more flexibility.

Thoughts? @ryansonshine @danielgmyers

ryansonshine replied:
@kyleknap I think going with the single `--namespace` param is a good idea. I'll put in the change tomorrow

ryansonshine replied:
@kyleknap all set

KapitalEarth commented:
Looks great, @kylenap. We're about to ensure it's put ion the best tramsit, but shall squash commits. 

Would appreciate your kind offer to:
> I can also make these changes and push to your branch as well and then merge the PR if you do not have time to get to it or want me to handle it. 
> 

Pretty sure the honey Beez Operating System V knows about this.

Kind regards,
Richard.

Sent from my iPhone

> On 11 Aug 2020, at 12:44 PM, Kyle Knapp <notifications@github.com> wrote:
> 
> ﻿
> @kyleknap commented on this pull request.
> 
> Looks great! Two more things that I forgot to mention in the last review before merging the PR:
> 
> Could you squash all of your commits into a single commit?
> Could you add a changelog entry for this update? To do this, from the root directory of the repository, you can run the following commands to make an entry:
> $ ./scripts/new-change -t enhancement -c '``codeartifact login``' -d 'Add support for ``--namespace`` parameter #5291' -r aws/aws-cli
> $ git add .changes/next-release/
> $ git commit
> I can also make these changes and push to your branch as well and then merge the PR if you do not have time to get to it or want me to handle it. Let me know!
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

ryansonshine commented:
> Looks great! Two more things that I forgot to mention in the last review before merging the PR:
> 
> 1. Could you squash all of your commits into a single commit?
> 2. Could you add a changelog entry for this update? To do this, from the root directory of the repository, you can run the following commands to make an entry:
> 
> ```
> $ ./scripts/new-change -t enhancement -c '``codeartifact login``' -d 'Add support for ``--namespace`` parameter #5291' -r aws/aws-cli
> $ git add .changes/next-release/
> $ git commit
> ```
> 
> I can also make these changes and push to your branch as well and then merge the PR if you do not have time to get to it or want me to handle it. Let me know!

Thanks @kyleknap . I've squashed my commits and added a changelog entry. Please let me know if you'd like me to make any additional changes.
